appscale/tools/agents/azure_agent.py

```diff
@@ -717,78 +793,156 @@ def terminate_instances(self, parameters):
-        for vm in vm_list:
-          if vm.name in instances_to_delete:
-            instances_to_delete.remove(vm.name)
-            thread = threading.Thread(target=self.delete_vmss_instance,
-                                      args=(compute_client, parameters,
-                                            vmss.name, vm.instance_id))
-            thread.start()
-            vmss_vm_delete_threads.append(thread)
-
-      for delete_thread in vmss_vm_delete_threads:
-        delete_thread.join()
+      vmss_vm_delete_futures = []
+      vmss_vm_delete_exceptions = set()
+
+      try:
+        vmss_vms_to_delete = [(vm.name, vmss.name) for vmss in vmss_list
+            for vm in compute_client.virtual_machine_scale_set_vms.list(
+            resource_group, vmss.name)
+            if vm.name in instances_to_delete]
+      except CloudError as e:
+        logging.exception("CloudError received while trying to terminate "
+                          "instances.")
+        raise AgentRuntimeException(e.message)
+      except ClientException as e:
+        logging.exception("ClientException received while attempting to "
+                          "contact Azure.")
+        raise AgentRuntimeException(e.message)
+
+      with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
+        for vm_name, vmss_name in vmss_vms_to_delete:
+          vmss_vm_delete_futures.append(executor.submit(
+              self.delete_vmss_instance, compute_client, parameters,
+              vmss_name, vm_name))
+          instances_to_delete.remove(vm_name)
+
+        for future in concurrent.futures.as_completed(vmss_vm_delete_futures):
+          exception = future.exception()
+          if exception:
+            vmss_vm_delete_exceptions.add(exception)
+
+      for exception in vmss_vm_delete_exceptions:
+        if not isinstance(exception, (CloudError, AgentRuntimeException)):
+          logging.exception(exception)
+      if vmss_vm_delete_exceptions:
+        raise AgentRuntimeException(json.dumps([str(
+            exception) for exception in vmss_vm_delete_exceptions]))
 
       AppScaleLogger.log("Virtual machine(s) have been successfully downscaled.")
       AppScaleLogger.log("Cleaning up any Scale Sets, if needed ...")
-      vmss_delete_threads = []
-      for vmss in vmss_list:
-        vm_list = compute_client.virtual_machine_scale_set_vms.list(
-          resource_group, vmss.name)
-        if not any(True for _ in vm_list):
-          thread = threading.Thread(
-            target=self.delete_virtual_machine_scale_set, args=(
-              compute_client, parameters, vmss.name))
-          thread.start()
-          vmss_delete_threads.append(thread)
-
-      for delete_thread in vmss_delete_threads:
-        delete_thread.join()
+      vmss_delete_futures = []
+      vmss_delete_exceptions = set()
+
+      try:
+        ss_to_delete = [vmss.name for vmss in vmss_list if
+                       sum(1 for _ in
```

Comment:
```
[vmss.name for vmss in vmss_list
 if not list(compute_client.virtual_machine_scale_set_vms.list(resource_group, vmss.name))]
```
is probably a bit simpler to read.

Reply:
s/list/any/
`if not any(..) ` might perform better, doesn't have to create a list.